protocols/request-response/tests/error_reporting.rs

```diff
@@ -0,0 +1,577 @@
+        wait_no_events(&mut swarm1).await;
+    };
+
+    // Client
+    //
+    // Expects OutboundFailure::Io failure with `FailOnReadResponse` error
+    let swarm2_task = async move {
+        let req_id = swarm2
+            .behaviour_mut()
+            .send_request(&peer1_id, Action::FailOnReadResponse);
+
+        let (peer, req_id_done, error) = wait_outbound_failure(&mut swarm2).await.unwrap();
+        assert_eq!(peer, peer1_id);
+        assert_eq!(req_id_done, req_id);
+
+        let error = match error {
+            OutboundFailure::Io(e) => e,
+            e => panic!("Unexpected error: {e:?}"),
+        };
+
+        assert_eq!(error.kind(), io::ErrorKind::Other);
+        assert_eq!(
+            error.into_inner().unwrap().to_string(),
+            "FailOnReadResponse"
+        );
+    };
+
+    let swarm1_task = pin!(swarm1_task);
+    let swarm2_task = pin!(swarm2_task);
+    futures::future::select(swarm1_task, swarm2_task).await;
+}
+
+#[async_std::test]
+async fn report_outbound_failure_on_write_request() {
+    let _ = env_logger::try_init();
+
+    let (peer1_id, mut swarm1) = new_swarm();
+    let (_peer2_id, mut swarm2) = new_swarm();
+
+    swarm1.listen().await;
+    swarm2.connect(&mut swarm1).await;
+
+    // Server
+    //
+    // Expects no events because `Event::Request` is produced after `read_request`.
+    let swarm1_task = async move {
+        // Keep the connection alive, otherwise swarm2 may receive `ConnectionClosed` instead
+        wait_no_events(&mut swarm1).await;
+    };
+
+    // Client
+    //
+    // Expects OutboundFailure::Io failure with `FailOnWriteRequest` error.
+    let swarm2_task = async move {
+        let req_id = swarm2
+            .behaviour_mut()
+            .send_request(&peer1_id, Action::FailOnWriteRequest);
+
+        let (peer, req_id_done, error) = wait_outbound_failure(&mut swarm2).await.unwrap();
+        assert_eq!(peer, peer1_id);
+        assert_eq!(req_id_done, req_id);
+
+        let error = match error {
+            OutboundFailure::Io(e) => e,
+            e => panic!("Unexpected error: {e:?}"),
+        };
+
+        assert_eq!(error.kind(), io::ErrorKind::Other);
+        assert_eq!(
+            error.into_inner().unwrap().to_string(),
+            "FailOnWriteRequest"
+        );
+    };
+
+    let swarm1_task = pin!(swarm1_task);
+    let swarm2_task = pin!(swarm2_task);
+    futures::future::select(swarm1_task, swarm2_task).await;
+}
+
+#[async_std::test]
+async fn report_outbound_timeout_on_read_response() {
+    let _ = env_logger::try_init();
+
+    // `swarm1` needs to have a bigger timeout to avoid racing
+    let (peer1_id, mut swarm1) = new_swarm_with_timeout(Duration::from_millis(200));
+    let (peer2_id, mut swarm2) = new_swarm_with_timeout(Duration::from_millis(100));
+
+    swarm1.listen().await;
+    swarm2.connect(&mut swarm1).await;
+
+    // Server
+    let swarm1_task = async move {
```

Comment:
If you do rename `swarm1` to `server`, you could spare the comments :)

Reply:
done